In [424]:
# Homework 2. Daniil Morozkov. 09/05/2023.
# import libraries
import pandas as pd
import numpy as np

Question 1
Using the full SPY dataset, compute daily close-to-close returns on SPY.  What is the average return on SPY over the full sample, during the first trading day of the month, and during the third-to-the-last trading day of the month?


In [425]:
# upload dataset, show few entries
df_spy = pd.read_csv('SPY.csv', parse_dates=[0], index_col=[0])
df_spy.head()

Open      High       Low     Close  Adj Close   Volume
Date                                                                  
1993-01-29  43.96875  43.96875  43.75000  43.93750  25.438095  1003200
1993-02-01  43.96875  44.25000  43.96875  44.25000  25.619024   480500
1993-02-02  44.21875  44.37500  44.12500  44.34375  25.673296   201300
1993-02-03  44.40625  44.84375  44.37500  44.81250  25.944683   529400
1993-02-04  44.96875  45.09375  44.46875  45.00000  26.053244   531500

In [426]:
# find returns
df_spy['Return'] = df_spy['Adj Close'].pct_change()
df_spy['Return']

Date
1993-01-29         NaN
1993-02-01    0.007113
1993-02-02    0.002118
1993-02-03    0.010571
1993-02-04    0.004184
                ...   
2022-07-20    0.006373
2022-07-21    0.010183
2022-07-22   -0.009278
2022-07-25    0.001215
2022-07-26   -0.011831
Name: Return, Length: 7426, dtype: float64

In [427]:
# drop na's
df_spy.dropna(inplace = True)

In [428]:
# find return over the sample
df_spy['Return'].mean()

0.0004385772725478149

In [429]:
# specify month and year for each entry
df_spy['Year'], df_spy['Month'] = df_spy.index.year, df_spy.index.month

In [430]:
# group by year, month and find mean return of the first day
df_spy.groupby(['Year', 'Month']).apply(lambda x: x.iloc[0]).mean()

Open         1.582197e+02
High         1.592608e+02
Low          1.570921e+02
Close        1.583282e+02
Adj Close    1.318684e+02
Volume       8.746886e+07
Return       2.441323e-03
Year         2.007336e+03
Month        6.466102e+00
dtype: float64

In [431]:
# group by year, month and find mean return of the third-to-the-last trading day of the month
df_grouped = df_spy.groupby(['Year', 'Month']).apply(lambda x: x.iloc[-3]).mean()
df_grouped, df_grouped['Return']

(Open         1.589775e+02
 High         1.599253e+02
 Low          1.578749e+02
 Close        1.589248e+02
 Adj Close    1.325908e+02
 Volume       8.047841e+07
 Return       6.807369e-04
 Year         2.007336e+03
 Month        6.466102e+00
 dtype: float64,
 0.0006807369336083447)

Question 2
Create a new dataframe that includes all dates on which there is price data available for SPY.  Add the T-Bill yield (the Close column of ^IRX) to this dataframe.  Note that ^IRX is missing on a number of dates that are in SPY, such as 2016-10-10.  You can fill in these missing values using the following command:
spy['Tbill yield filled'] = spy['Tbill yield'].fillna(method='ffill')
You should take a look at how this works and think about why it is probably ok.  
Compute T-bill returns following the instructions at the bottom of this sheet.  What are the mean and standard deviation of excess SPY returns, defined as the SPY return minus the T-bill return?

In [432]:
# load datasets
irx = pd.read_csv('^IRX.csv',  parse_dates=[0], index_col=[0])
spy = pd.read_csv('SPY.csv', parse_dates=[0], index_col=[0])
irx.info(), spy.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 19450 entries, 1960-01-04 to 2022-07-26
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Open       15651 non-null  float64
 1   High       15651 non-null  float64
 2   Low        15651 non-null  float64
 3   Close      15651 non-null  float64
 4   Adj Close  15651 non-null  float64
 5   Volume     15651 non-null  float64
dtypes: float64(6)
memory usage: 1.0 MB
<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 7426 entries, 1993-01-29 to 2022-07-26
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Open       7426 non-null   float64
 1   High       7426 non-null   float64
 2   Low        7426 non-null   float64
 3   Close      7426 non-null   float64
 4   Adj Close  7426 non-null   float64
 5   Volume     7426 non-null   int64  
dtypes: float64(5), int64(1)
memory usage: 406.1 KB


(None, None)

In [433]:
# add t-bill yield
spy['Tbill yield'] = irx['Close']
spy

Open        High         Low       Close   Adj Close  \
Date                                                                     
1993-01-29   43.968750   43.968750   43.750000   43.937500   25.438095   
1993-02-01   43.968750   44.250000   43.968750   44.250000   25.619024   
1993-02-02   44.218750   44.375000   44.125000   44.343750   25.673296   
1993-02-03   44.406250   44.843750   44.375000   44.812500   25.944683   
1993-02-04   44.968750   45.093750   44.468750   45.000000   26.053244   
...                ...         ...         ...         ...         ...   
2022-07-20  392.470001  396.260010  391.029999  394.769989  394.769989   
2022-07-21  394.160004  398.839996  391.630005  398.790009  398.790009   
2022-07-22  398.920013  400.179993  392.750000  395.089996  395.089996   
2022-07-25  395.750000  396.470001  393.209991  395.570007  395.570007   
2022-07-26  393.839996  394.059998  389.950012  390.890015  390.890015   

              Volume  Tbill yield  
Date                               
1993-01-29   1003200        2.900  
1993-02-01    480500        2.900  
1993-02-02    201300        2.960  
1993-02-03    529400        2.930  
1993-02-04    531500        2.900  
...              ...          ...  
2022-07-20  71843800        2.378  
2022-07-21  64903900        2.360  
2022-07-22  72130300        2.355  
2022-07-25  53631500        2.393  
2022-07-26  52642500        2.475  

[7426 rows x 7 columns]

In [434]:
# fill na's
spy['Tbill yield filled'] = spy['Tbill yield'].fillna(method='ffill')

In [435]:
# looks ok
spy['Tbill yield filled'].loc[spy.index=='2016-10-10']

Date
2016-10-10    0.31
Name: Tbill yield filled, dtype: float64

In [436]:
# compute Nt by checking difference between dates
spy['Nt'] = spy.index.to_series().diff().dt.days.fillna(1)
print(spy['Nt'])

Date
1993-01-29    1.0
1993-02-01    3.0
1993-02-02    1.0
1993-02-03    1.0
1993-02-04    1.0
             ... 
2022-07-20    1.0
2022-07-21    1.0
2022-07-22    1.0
2022-07-25    3.0
2022-07-26    1.0
Name: Nt, Length: 7426, dtype: float64


In [437]:
# compute risk-free return
spy['rf'] = spy['Tbill yield filled'].shift(1) * spy['Nt']/36500
spy['rf']

Date
1993-01-29         NaN
1993-02-01    0.000238
1993-02-02    0.000079
1993-02-03    0.000081
1993-02-04    0.000080
                ...   
2022-07-20    0.000067
2022-07-21    0.000065
2022-07-22    0.000065
2022-07-25    0.000194
2022-07-26    0.000066
Name: rf, Length: 7426, dtype: float64

In [438]:
# find spy returns
spy['Return'] = spy['Adj Close'].pct_change()
spy['Return']

Date
1993-01-29         NaN
1993-02-01    0.007113
1993-02-02    0.002118
1993-02-03    0.010571
1993-02-04    0.004184
                ...   
2022-07-20    0.006373
2022-07-21    0.010183
2022-07-22   -0.009278
2022-07-25    0.001215
2022-07-26   -0.011831
Name: Return, Length: 7426, dtype: float64

In [439]:
# find excess returns mean and SD
spy['Excess Returns'] = spy['Return'] - spy['rf']
spy.dropna(inplace=True)
spy['Excess Returns'].describe()

count    7403.000000
mean        0.000351
std         0.011886
min        -0.109444
25%        -0.004468
50%         0.000604
75%         0.005860
max         0.145180
Name: Excess Returns, dtype: float64

Question 3
What is the maximum drawdown of OAKMX?  (Hint: use adjusted closing prices for portfolio values and then the cummax() method to get the prior highs.)

In [440]:
# load dataset
oakmx = pd.read_csv('OAKMX.csv',  parse_dates=[0], index_col=[0])

In [441]:
# the cummax() method to get the prior highs
oakmx['Prior Max'] = oakmx['Adj Close'].cummax()

In [442]:
# calculate drawdowns
oakmx['Drawdowns'] = (oakmx['Adj Close'] - oakmx['Prior Max'])/oakmx['Prior Max']

In [443]:
# find min value for drowdowns (largest drawdown)
oakmx['Drawdowns'].min()

-0.56194799486071

Question 4
What is the downside deviation of OAKMX?  As a benchmark return, use the T-Bill return.


In [444]:
# calculate oakmax return
oakmx['Return'] = oakmx['Adj Close'].pct_change()
oakmx['Return'].dropna(inplace=True)

In [445]:
# find excces return for oakmx
oakmx['Excess Return'] = oakmx['Return'] - spy['rf']
stats = oakmx['Excess Return'].describe()
stats

count    7403.000000
mean        0.000364
std         0.011548
min        -0.131806
25%        -0.004426
50%         0.000670
75%         0.005460
max         0.107211
Name: Excess Return, dtype: float64

In [446]:
# find downside deviation
oakmx['min squared'] = 0
oakmx['min squared'].loc[oakmx['Excess Return']<0] = oakmx['Excess Return']**2
min2stats = oakmx['min squared'].describe()
stats.loc['Downside Deviation'] = np.sqrt(min2stats.loc['mean'])
print(stats)

count                 7403.000000
mean                     0.000364
std                      0.011548
min                     -0.131806
25%                     -0.004426
50%                      0.000670
75%                      0.005460
max                      0.107211
Downside Deviation       0.008011
Name: Excess Return, dtype: float64


C:\Users\moroz\AppData\Local\Temp\ipykernel_2780\1687295211.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  oakmx['min squared'].loc[oakmx['Excess Return']<0] = oakmx['Excess Return']**2


Question 5
What is the average excess return on OAKMX when the previous Adjusted Closing price is above its 200-day moving average?  What about when it is below the moving average?  (Hint: use the rolling() method to get the moving average.)  Are these means statistically significant?


In [447]:
# find moving average 200 for oakmx
oakmx['MA_200'] = oakmx['Adj Close'].rolling(window=200).mean()
oakmx.dropna(inplace=True)

In [448]:
# find cases when returns are greate
oakmx['Prev_MA_200'] = oakmx['MA_200'].shift(1)
oakmx['Prev_Adj Close'] = oakmx['Adj Close'].shift(1)
above_ma = oakmx[oakmx['Prev_Adj Close'] > oakmx['Prev_MA_200']]['Excess Return']
below_ma = oakmx[oakmx['Prev_Adj Close'] <= oakmx['Prev_MA_200']]['Excess Return']


In [451]:
# What is the average excess return on OAKMX when the previous Adjusted Closing price is above its 200-day moving average?  What about when it is below the moving average?  
above_ma.mean(), below_ma.mean()

(0.0003938579981215834, 0.00026732878718344435)

In [452]:
def calc_t_stat(df):
    sample_mean = df.mean()
    sample_std = df.std()
    n = len(df)
    t_stat = sample_mean / (sample_std / n ** (1/2))
    return t_stat

calc_t_stat(above_ma), calc_t_stat(below_ma)

(3.649496029408021, 0.5901956107225281)

At significance level alpha = 0.05, t crit = 1.96, we may conclude that above_ma mean excess return is statistically signficant (reject H0, 3.64>1.96), below_ma is not significant (accept H0 0.59<1.96)